src/upload/index.ts

```diff
@@ -0,0 +1,201 @@
+import {
+  HeadObjectCommand,
+  NotFound,
+  PutObjectCommand,
+  S3Client,
+} from "@aws-sdk/client-s3";
+import * as zlib from "zlib";
+import fs from "fs";
+
+import { LightBlockCache, lightBlockCache } from "@/cache";
+import { LightBlock } from "@/models/lightstreamer";
+import { logger } from "@/utils/logger";
+
+class UploadError extends Error {}
+
+export type BlockRange = {
+  start: number;
+  end: number;
+};
+
+export type BlockFile = {
+  file: string;
+  manifest: string;
+};
+
+export class LightBlockUpload {
+  private cache: LightBlockCache;
+  private s3Client: S3Client;
+  private chunkSizeMb: number;
+  private bucket: string;
+
+  constructor(cache: LightBlockCache) {
+    this.cache = cache;
+    if (!process.env["BUCKET_ENDPOINT"]) {
+      throw new UploadError("BUCKET_ENDPOINT not set");
+    }
+    if (!process.env["BUCKET_ACCESS_KEY_ID"]) {
+      throw new UploadError("BUCKET_ACCESS_KEY_ID not set");
+    }
+    if (!process.env["BUCKET_SECRET_ACCESS_KEY"]) {
+      throw new UploadError("BUCKET_SECRET_ACCESS_KEY not set");
+    }
+    if (!process.env["BUCKET_NAME"]) {
+      throw new UploadError("BUCKET_NAME not set");
+    }
+    if (!process.env["UPLOAD_CHUNK_SIZE_MB"]) {
+      throw new UploadError("UPLOAD_CHUNK_SIZE_MB not set");
+    }
+    this.chunkSizeMb = parseInt(
+      process.env["UPLOAD_CHUNK_SIZE_MB"] as string,
+      10,
+    );
+    this.bucket = process.env["BUCKET_NAME"];
+    this.s3Client = new S3Client({
+      region: "auto",
+      endpoint: process.env["BUCKET_ENDPOINT"],
+      credentials: {
+        accessKeyId: process.env["BUCKET_ACCESS_KEY_ID"],
+        secretAccessKey: process.env["BUCKET_SECRET_ACCESS_KEY"],
+      },
+    });
+  }
+
+  async upload(): Promise<void> {
+    // eslint-disable-next-line no-constant-condition
+    const fileName = "blocks";
+    const currentUploadSize = (await this.getFileSize(fileName)) || 0;
+    logger.info(
+      `Current file uploaded size: ${currentUploadSize}, creating new upload...`,
+    );
+    const files = await this.createBlockFiles(fileName, currentUploadSize);
+
+    logger.info(`Upload: begin...`);
+    const gzip = await this.gzipFile(files.file, `${files.file}.gz`);
+    await this.uploadFile(gzip, "application/gzip");
+    logger.info(`Upload: gzip file complete`);
+
+    await this.uploadFile(files.file, "application/octet-stream");
+    logger.info(`Upload: binary file complete`);
+
+    const gzipManifest = await this.gzipFile(
+      files.manifest,
+      `${files.manifest}.gz`,
+    );
+    await this.uploadFile(gzipManifest, "application/gzip");
+    logger.info(`Upload: manifest file complete`);
```

Comment:
These files don't look like they're versioned, so while this method is running, there will be a period of time where `blocks`, `blocks.gz`, and `blocks.manifest.gz` will have inconsistent contents.

In other words, if I have a client and I download these 3 files at a random time, it's possible that these files will have different contents. There's no way for me to e.g. download a manifest, inspect that manifest, and then download the compressed blocks that correspond to that manifest, because by the time I download the blocks, it's possible that they might have changed.

Another, related, issue is that if an error occurs in any of these `uploadFile` calls, then the other files won't be uploaded, ever.

Reply:
This is a good idea. Perhaps I can upload the files to s staging directory, and if all three succeed move them all to the download link location? I could also wrap this logic in a failure handler that retries if there are any errors.

Reply:
That would improve the situation a lot, but still allows a short time window in which the files will be inconsistent. Something else to consider is: what happens if a file on R2 gets removed/overwritten while someone is downloading it? Will their download continue undisrupted, will it stop, or will it contain mixed data?

IMO the best thing to do would be to: place the files in a timestamped directory, and have a file at the root with the name of the latest directory. This would kind of work like a symlink. The client would have to check what is the latest directory and then download from there. As long as we keep a time window large enough between directory creation and deletion, then clients shouldn't be affected in any negative way. There's one more file to download though, but that's a small price to pay.

Reply:
Thanks for the great suggestion, I went with this approach.